portal/api-authentication-grpc.go

```diff
@@ -19,12 +19,94 @@ package portal
 import (
 	"context"
 
+	"database/sql"
+	"fmt"
+
 	"golang.org/x/crypto/bcrypt"
 
 	cluster "github.com/minio/m3/cluster"
 	pb "github.com/minio/m3/portal/stubs"
+	"github.com/minio/minio-go/v6"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
 )
 
+type User struct {
+	Tenant   string
+	Email    string
+	IsAdmin  bool
+	Password string
+	UUID     string
+}
+
+func (s *server) MakeBucket(ctx context.Context, in *pb.MakeBucketRequest) (res *pb.Bucket, err error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return nil, status.New(codes.InvalidArgument, "Metadata not found").Err()
+	}
+
+	var sessionId string
+	switch sIds := md.Get("sessionId"); len(sIds) {
+	case 0:
+		return nil, status.New(codes.Unauthenticated, "SessionId not found").Err()
+	default:
+		sessionId = sIds[0]
+	}
+
+	// With validating sessionId behind us, we query the tenant MinIO
+	// service corresponding to the logged-in user to make the bucket
+
+	// Prepare DB instance
+	db := cluster.GetInstance().Db
+	// Get tenant name from the DB
+	getTenantShortnameQ := `SELECT t.short_name
+                           FROM m3.provisioning.sessions as s JOIN m3.provisioning.tenants as t
+                           ON (s.tenant_id = t.id) WHERE s.id=$1`
+	tenantRow := db.QueryRow(getTenantShortnameQ, sessionId)
+
+	var tenantShortname string
+	err = tenantRow.Scan(&tenantShortname)
+	if err == sql.ErrNoRows {
+		return nil, status.New(codes.NotFound, "No matching session found").Err()
+	}
+	if err != nil {
+		return nil, status.New(codes.InvalidArgument, err.Error()).Err()
+	}
+
+	appCtx, _ := cluster.NewContext(tenantShortname)
+
+	// Get storage group where the tenant is present
+	sgt := <-cluster.GetTenantStorageGroupByShortName(appCtx, tenantShortname)
+
+	// Get the credentials for a tenantShortName
+	var tenantConf *cluster.TenantConfiguration
+	tenantConf, err = cluster.GetTenantConfig(sgt.Tenant.ShortName)
+	if err != nil {
+		return nil, status.New(codes.NotFound, "Tenant config missing").Err()
+	}
+
+	// Build tenant MinIO service endpoint
+	tenantAddress := fmt.Sprintf("%s:%d", sgt.ServiceName, sgt.Port)
+	// Initialize minio client object.
+	minioClient, err := minio.New(tenantAddress,
+		tenantConf.AccessKey,
+		tenantConf.SecretKey,
+		false)
+	if err != nil {
+		return nil, status.New(codes.Internal, "Failed to connect to object storage").Err()
+	}
+
+	// Create bucket in the tenant's MinIO
+	// TODO: Need to decide if we are going to support regions
+	bucket := in.GetName()
```

Comment:
bucket name needs validation, or do you want MinIO to error out and use that as validation?

Reply:
I would rely on Minio's validation unless m3 has more restrictions on bucket names.

Reply:
that error is because you need to port-forward the service to your local and add it to `/etc/hosts` alternatively you can use `kubefwd` so you local can reach the desired MinIO